cli/azd/cmd/container.go

```diff
@@ -534,6 +585,29 @@ func registerCommonDependencies(container *ioc.NestedContainer) {
 		}
 	}
 
+	// Platforms
+	platformProviderMap := map[project.PlatformKind]any{
+		devcenter.PlatformKindDevCenter: devcenter.NewPlatform,
+	}
+
+	for provider, constructor := range platformProviderMap {
+		platformName := fmt.Sprintf("%s-platform", provider)
+		if err := container.RegisterNamedSingleton(platformName, constructor); err != nil {
+			panic(fmt.Errorf("registering platform provider %s: %w", provider, err))
+		}
+
+		var platform project.PlatformProvider
+		if err := container.ResolveNamed(platformName, &platform); err != nil {
+			panic(fmt.Errorf("resolving platform provider %s: %w", provider, err))
+		}
+
+		if platform.IsEnabled() {
+			if err := platform.ConfigureContainer(container); err != nil {
```

Comment:
This would be a good topic for our Engineering chats to walk everyone through the platform provider feature that I'm introducing here.  Think of the `container.go` as the default container configuration for all azd.  Then a platform provider has its own `ConfigureContainer` implementation where it can add/change container configuration to enable/disable/default specific things like template sources, provision providers, etc.

Reply:
Is this a common pattern? (something we can reference as the general strategy we are following?  or some other system that implements container + platform, where the platform is a way to mutate its container?  Is this similar to the hierarchy + abstract-class pattern, where a subclass can override the base class with its own implementation?

I'm just wondering if there might be some idiomatic way to solve this problem with go.

Reply:
I'm not sure about Go, but this is a very common pattern in other pluggable systems like .NET web apps.  When you register a new feature part of the registration process adds/modifies container registrations to support DI throughout the rest of the platform.

Reply:
I think there's not an easy way to know what's happening here.

The concept of the `platform` is working as an override for the previous registrations, without no clue for a reader to discover about how this is working.

Can we move to a world where we have an `azd platform` and `ade platform` and there's a way we select either one of the other?  That would make it clear that a platform is mutually exclusive (can't keep applying platforms on top of other platforms) and removes the fragility of depending on the ordering of the registrations (registering something after the platform would override the platform settings and break it)

Reply:
Default `azd` behavior is our implicit platform and everything contained within the `container.go`.  Anything split out into a `PlatformProvider` is supplemental to our defaults and could extend or override behavior.

Platform registrations always come after our defaults.  Anything that breaks should be caught in tests and I would consider it to be more of a dev time error vs any user error.